Fixes #168: Don't pass -a by default when doing a simple commit().

geerlingguy opened the pull request.

GrahamCampbell commented:
This is a breaking change, and would break my code if it were merged.

geerlingguy commented:
> This is a breaking change, and would break my code if it were merged.

Most definitely, which is why if this were merged it should at minimum be a minor version bump... but probably reserved for a major version bump.

See #168 for details. The idea behind the PR is to make this git library behave more like Git itself—when you just do a plain commit with a message, it doesn't also add all modified files. (That really confused me the first time it happened... then I also realized that you have to use a different syntax for a commit to get it to _not_ pass `-a`, which was even more confusing).

GrahamCampbell commented:
Why not just add a new method that behaves in the new way, to avoid breaking existing code.

GrahamCampbell commented:
> Most definitely, which is why if this were merged it should at minimum be a minor version bump... but probably reserved for a major version bump.

If we're following semver, this would need to be 3.0.0. If we want to put this change in a new major version, I'm happy with it's current form. If not, we shouldn't be breaking 2.x.

geerlingguy commented:
> Why not just add a new method that behaves in the new way, to avoid breaking existing code.

Because the current method, `commit()`, is currently a bit overreaching.

If I added a new method, I would have to do something like `commitWithoutAll()`, and it would be kind of weird because if you come from using `git`, and want to use this library to wrap `git`, then your expectation would be that a bare `commit()` would be the same as (or similar to) `git commit`.

But in this case, the way it _currently_ works, it should probably be called `commitAll()` or something like that, because it's adding in a `-a`, and by documentation, by method name, and by convention, that seems unintuitive to me.

I've changed my code to do the workaround to explicitly set the message using an array, but as a convenience, it was tripping me up for a couple hours that `commit()` with a message was adding all modified files. I thought I was going insane.

TomasVotruba commented:
I there would be a new method, I can merge it right now

valeryan commented:
Looking at this command it seems that it is already making assumptions by the number of arguments passed. Why not just add a logic block that would assume that if an argument of `a` is passed in the array like `$git->commit('Makefile', ['m' => 'My commit message', 'a' => false]);` that it overrides the flag. This does not explicitly change the API of the method and would not break existing code. It still does not make the default behavior work like `git commit` but it would allow the behavior to be configurable.

TomasVotruba commented:
Closing as outdated
